diff --git a/arch/arm/dts/sun8i-h3-nanopi-m1.dts b/arch/arm/dts/sun8i-h3-nanopi-m1.dts
index ec63d10..652a4dd 100644
--- a/arch/arm/dts/sun8i-h3-nanopi-m1.dts
+++ b/arch/arm/dts/sun8i-h3-nanopi-m1.dts
@@ -47,6 +47,30 @@
 	compatible = "friendlyarm,nanopi-m1", "allwinner,sun8i-h3";
 };
 
+&emac {
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	allwinner,leds-active-low;
+	status = "okay";
+};
+
+&mmc2 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&mmc2_8bit_pins>;
+        vmmc-supply = <&reg_vcc3v3>;
+        bus-width = <8>;
+        non-removable;
+        status = "okay";
+        cap-mmc-hw-reset;
+};
+
+&mmc2_8bit_pins {
+        /* Increase drive strength for DDR modes */
+        allwinner,drive = <SUN4I_PINCTRL_40_MA>;
+        /* eMMC is missing pull-ups */
+        allwinner,pull = <SUN4I_PINCTRL_PULL_UP>;
+};
+
 &ehci1 {
 	status = "okay";
 };
diff --git a/arch/arm/dts/sun8i-h3-qpt-vdiv1.dts b/arch/arm/dts/sun8i-h3-qpt-vdiv1.dts
new file mode 100644
index 0000000..4ba3a89
--- /dev/null
+++ b/arch/arm/dts/sun8i-h3-qpt-vdiv1.dts
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2016 Milo Kim <woogyom.kim@gmail.com>
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPL or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This file is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This file is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "sun8i-h3-nanopi.dtsi"
+
+/ {
+	model = "QPT VDIH3v1";
+	compatible = "QPT,VDIH3v1", "allwinner,sun8i-h3";
+
+        leds {
+		compatible = "gpio-leds";
+
+		led2 {
+			label = "QPT:LED2"; /* PA15 */
+			gpios = <&pio 0 15 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		led3 {
+			label = "QPT:LED3"; /* PE0 */
+			gpios = <&pio 4 0 GPIO_ACTIVE_LOW>;
+		};
+
+		led4 {
+			label = "QPT:LED4"; /* PE1 */
+			gpios = <&pio 4 1 GPIO_ACTIVE_LOW>;
+		};
+
+		led5 {
+			label = "QPT:LED5"; /* PE11 */
+			gpios = <&pio 4 11 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&emac {
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	allwinner,leds-active-low;
+	status = "okay";
+};
+
+&mmc2 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&mmc2_8bit_pins>;
+        vmmc-supply = <&reg_vcc3v3>;
+        bus-width = <8>;
+        non-removable;
+        status = "okay";
+        cap-mmc-hw-reset;
+};
+
+&mmc2_8bit_pins {
+        /* Increase drive strength for DDR modes */
+        allwinner,drive = <SUN4I_PINCTRL_40_MA>;
+        /* eMMC is missing pull-ups */
+        allwinner,pull = <SUN4I_PINCTRL_PULL_UP>;
+};
+
+&ehci1 {
+	status = "okay";
+};
+
+&ehci2 {
+	status = "okay";
+};
+
+&ohci1 {
+	status = "okay";
+};
+
+&ohci2 {
+	status = "okay";
+};
diff --git a/cmd/usb.c b/cmd/usb.c
index 0ccb1b5..424dbd0 100644
--- a/cmd/usb.c
+++ b/cmd/usb.c
@@ -583,7 +583,7 @@ static void do_usb_start(void)
 {
 	bootstage_mark_name(BOOTSTAGE_ID_USB_START, "usb_start");
 
-	if (usb_init() < 0)
+	if (0 && usb_init() < 0) /* ROY: disable usb */
 		return;
 
 	/* Driver model will probe the devices as they are found */
diff --git a/common/autoboot.c b/common/autoboot.c
index 9b9fb53..52361be 100644
--- a/common/autoboot.c
+++ b/common/autoboot.c
@@ -13,6 +13,8 @@
 #include <menu.h>
 #include <post.h>
 #include <u-boot/sha256.h>
+#include <asm/gpio.h>
+#include <fs.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -27,6 +29,131 @@ DECLARE_GLOBAL_DATA_PTR;
 /* Stored value of bootdelay, used by autoboot_command() */
 static int stored_bootdelay;
 
+#define GPIO_RESET_KEY "PL3"
+
+typedef enum {
+	EMC_RST_NOT_SUPPORTED = -1, 	//hw not supported
+	EMC_NORMAL_BOOT = 0, 			//normal boot to system
+	EMC_KEY_PRESSED, 		//reset key pressed
+	EMC_SOFT_RECOVERY, 		//file flags touched
+	EMC_CONS_RECOVERY, 		//console key input
+} emergency_t;
+
+static emergency_t emergency_boot = EMC_RST_NOT_SUPPORTED;
+
+static int __emc_key_read(void)
+{
+	int n;
+	unsigned int gpio;
+
+	n = gpio_lookup_name(GPIO_RESET_KEY, NULL, NULL, &gpio);
+	if(n) {
+		printf("%s: GPIO not found %s\n", __FUNCTION__, GPIO_RESET_KEY);
+		return -1;
+	}
+
+	n = gpio_request(gpio, "boot-emergency");
+	if(n<0) {
+		printf("%s: request %s failed\n", __FUNCTION__, GPIO_RESET_KEY);
+		return -1;
+	}
+
+	//set dir input
+	gpio_direction_input(gpio);
+
+	//get value
+	n =	gpio_get_value(gpio); //PL3
+
+	gpio_free(gpio);
+	return n;
+}
+
+/*
+* HW version: 1.0, the PL3 default 0, and none key layout.
+* HW version: 1.1, the PL3 default 1, and RESET key attached.
+*/
+void emc_check_hw_support(void)
+{
+	if(__emc_key_read() > 0) {
+		/* only hw version >= 1.1 , RESET key supported and default = 1 */
+		emergency_boot = EMC_NORMAL_BOOT;
+		printf("%s: hw emergency boot supported.\n", __FUNCTION__);
+	}
+	return;
+}
+
+static void __mmc_recovery_flag(void)
+{
+	int n = file_exists("mmc", "1:3", "/.auto_recovery", FS_TYPE_ANY);
+	if(n>0) {
+		emergency_boot = EMC_SOFT_RECOVERY;
+		printf("%s: found mmc boot recovery flags\n", __FUNCTION__);
+	}
+}
+
+/* called many time as boot */
+static void __emergency_trigger(void)
+{
+	int n;
+
+	if(emergency_boot != EMC_NORMAL_BOOT)
+		return;
+
+	/* check hw reset key status */
+	n = __emc_key_read();
+	if(n >= 0) {
+		/* default: 1, PRESSED = 0 */
+		emergency_boot = (!n ? EMC_KEY_PRESSED : EMC_NORMAL_BOOT);
+	}
+
+	if(emergency_boot > EMC_NORMAL_BOOT)
+		printf("%s: key pressed %d\n", __FUNCTION__, emergency_boot);
+	return;
+}
+
+static void __emergency(void)
+{
+	char buff[128];
+
+	/*
+	* check gpio key status
+	*/
+	printf("%s: check if emergency.\n", __FUNCTION__);
+	if(emergency_boot <= EMC_NORMAL_BOOT) {
+		return;
+	}
+
+	/*
+		write 0x42000000 $BOARD/zImage 		//2048
+		write 0x43000000 $BOARD/script.bin 	//1024
+		write 0x43300000 $BOARD/uInitrd		//16348
+		write 0x43100000 $BOARD/boot.scr	//1152
+
+		rootfs.bin: 16M 0x8000@0xd000
+		script.bin: 1M@128K
+		boot-scr.bin: 1M + 128K@128K
+		kernel.bin: 2M + 0x1000@0x7000
+	*/
+	printf("Boot into emergency mode...\n");
+
+	snprintf(buff, sizeof(buff), "earlyprintk=ttyS0,115200 " \
+						"console=ttyS0,115200 " \
+						"console=tty1 " \
+						"panic=10 " \
+						"emergency_type=%u ", emergency_boot);
+
+	env_set("bootargs", buff);
+
+	env_set("emergency_boot", "mmc dev 1 && " \
+						"mmc read ${fdt_addr_r} 0x0800 0x100 && " \
+						"mmc read ${scriptaddr} 0x0900 0x100 && " \
+						"mmc read ${ramdisk_addr_r} 0x8000 0xd000 && " \
+						"mmc read ${kernel_addr_r} 0x1000 0x7000");
+
+	env_set("gogo", "run emergency_boot && source ${scriptaddr}");
+	return;
+}
+
 #if defined(CONFIG_AUTOBOOT_KEYED)
 #if defined(CONFIG_AUTOBOOT_STOP_STR_SHA256)
 
@@ -101,6 +228,7 @@ static int passwd_abort(uint64_t etime)
 #else
 static int passwd_abort(uint64_t etime)
 {
+	int blink = 0;
 	int abort = 0;
 	struct {
 		char *str;
@@ -171,6 +299,8 @@ static int passwd_abort(uint64_t etime)
 				abort = 1;
 			}
 		}
+		qpt_leds_blink(++blink / 2000);
+		__emergency_trigger();
 	} while (!abort && get_ticks() <= etime);
 
 	return abort;
@@ -266,6 +396,11 @@ static int abortboot(int bootdelay)
 		gd->flags &= ~GD_FLG_SILENT;
 #endif
 
+	if(abort) {
+		printf("ENTRY emergency boot manually...\n");
+		emergency_boot = EMC_CONS_RECOVERY;
+	}
+
 	return abort;
 }
 
@@ -342,7 +477,11 @@ void autoboot_command(const char *s)
 {
 	debug("### main_loop: bootcmd=\"%s\"\n", s ? s : "<UNDEFINED>");
 
+	/* check filesystem flags */
+	__mmc_recovery_flag();
+
 	if (stored_bootdelay != -1 && s && !abortboot(stored_bootdelay)) {
+		__emergency();
 #if defined(CONFIG_AUTOBOOT_KEYED) && !defined(CONFIG_AUTOBOOT_KEYED_CTRLC)
 		int prev = disable_ctrlc(1);	/* disable Control C checking */
 #endif
@@ -353,6 +492,8 @@ void autoboot_command(const char *s)
 		disable_ctrlc(prev);	/* restore Control C checking */
 #endif
 	}
+	__emergency();
+
 
 #ifdef CONFIG_MENUKEY
 	if (menukey == CONFIG_MENUKEY) {
diff --git a/common/usb_storage.c b/common/usb_storage.c
index 9cd6474..7fa1541 100644
--- a/common/usb_storage.c
+++ b/common/usb_storage.c
@@ -47,6 +47,24 @@
 #include <part.h>
 #include <usb.h>
 
+#include <led.h>
+void qpt_leds_blink(int i)
+{
+	struct udevice *dev;
+	char buff[16];
+	int res;
+
+	snprintf(buff, sizeof(buff), "QPT:LED%d", (i%4+2));
+	res = led_get_by_label(buff, &dev);
+	if(res) {
+		printf("%s: Failed to get %s\n", __func__, buff);
+		return;
+	}
+
+	res = (led_get_state(dev) == LEDST_OFF) ? LEDST_ON : LEDST_OFF;
+	led_set_state(dev, res);
+}
+
 #undef BBB_COMDAT_TRACE
 #undef BBB_XPORT_TRACE
 
@@ -279,6 +297,7 @@ static int usb_stor_probe_device(struct usb_device *udev)
 			part_init(blkdev);
 			debug("partype: %d\n", blkdev->part_type);
 			usb_max_devs++;
+			qpt_leds_blink(usb_max_devs);
 			debug("%s: Found device %p\n", __func__, udev);
 		}
 	}
diff --git a/config.tempe-nanopi-m1 b/config.tempe-nanopi-m1
index 466bbe0..63ccb33 100644
--- a/config.tempe-nanopi-m1
+++ b/config.tempe-nanopi-m1
@@ -194,7 +194,7 @@ CONFIG_MMC3_CD_PIN=""
 CONFIG_MMC1_PINS=""
 CONFIG_MMC2_PINS=""
 CONFIG_MMC3_PINS=""
-CONFIG_MMC_SUNXI_SLOT_EXTRA=-1
+CONFIG_MMC_SUNXI_SLOT_EXTRA=2
 CONFIG_INITIAL_USB_SCAN_DELAY=0
 CONFIG_USB0_VBUS_PIN=""
 CONFIG_USB0_VBUS_DET=""
@@ -231,15 +231,15 @@ CONFIG_ARMV7_PSCI_NR_CPUS=4
 # ARM debug
 #
 # CONFIG_DEBUG_LL is not set
-CONFIG_DEFAULT_DEVICE_TREE="sun8i-h3-nanopi-m1"
+CONFIG_DEFAULT_DEVICE_TREE="sun8i-h3-qpt-vdiv1"
 # CONFIG_DEBUG_UART is not set
 # CONFIG_AHCI is not set
 
 #
 # General setup
 #
-CONFIG_LOCALVERSION="-armbian"
-# CONFIG_LOCALVERSION_AUTO is not set
+CONFIG_LOCALVERSION=""
+CONFIG_LOCALVERSION_AUTO=y
 CONFIG_CC_OPTIMIZE_FOR_SIZE=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_ENV_VARS_UBOOT_CONFIG=y
@@ -288,7 +288,7 @@ CONFIG_BOOTSTAGE_STASH_SIZE=0x1000
 # CONFIG_SATA_BOOT is not set
 # CONFIG_SD_BOOT is not set
 # CONFIG_SPI_BOOT is not set
-CONFIG_BOOTDELAY=1
+CONFIG_BOOTDELAY=2
 # CONFIG_USE_BOOTARGS is not set
 CONFIG_USE_BOOTCOMMAND=y
 CONFIG_BOOTCOMMAND="run distro_bootcmd"
@@ -572,7 +572,7 @@ CONFIG_CMD_BMP=y
 # CONFIG_CMD_BKOPS_ENABLE is not set
 # CONFIG_CMD_CACHE is not set
 # CONFIG_CMD_DISPLAY is not set
-# CONFIG_CMD_LED is not set
+CONFIG_CMD_LED=y
 # CONFIG_CMD_DATE is not set
 # CONFIG_CMD_TIME is not set
 # CONFIG_CMD_GETTIME is not set
@@ -669,10 +669,9 @@ CONFIG_MKIMAGE_DTC_PATH="dtc"
 #
 # Environment
 #
-# CONFIG_ENV_IS_NOWHERE is not set
 # CONFIG_ENV_IS_IN_EEPROM is not set
-# CONFIG_ENV_IS_IN_FAT is not set
-CONFIG_ENV_IS_IN_EXT4=y
+CONFIG_ENV_IS_IN_FAT=y
+# CONFIG_ENV_IS_IN_EXT4 is not set
 # CONFIG_ENV_IS_IN_FLASH is not set
 # CONFIG_ENV_IS_IN_MMC is not set
 # CONFIG_ENV_IS_IN_NAND is not set
@@ -681,9 +680,9 @@ CONFIG_ENV_IS_IN_EXT4=y
 # CONFIG_ENV_IS_IN_REMOTE is not set
 # CONFIG_ENV_IS_IN_SPI_FLASH is not set
 # CONFIG_ENV_IS_IN_UBI is not set
-CONFIG_ENV_EXT4_INTERFACE="mmc"
-CONFIG_ENV_EXT4_DEVICE_AND_PART="0:auto"
-CONFIG_ENV_EXT4_FILE="/boot/boot.env"
+CONFIG_ENV_FAT_INTERFACE="mmc"
+CONFIG_ENV_FAT_DEVICE_AND_PART="1:auto"
+CONFIG_ENV_FAT_FILE="uboot.env"
 # CONFIG_USE_DEFAULT_ENV_FILE is not set
 # CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG is not set
 CONFIG_NET=y
@@ -818,7 +817,9 @@ CONFIG_DM_KEYBOARD=y
 #
 # LED Support
 #
-# CONFIG_LED is not set
+CONFIG_LED=y
+CONFIG_LED_BLINK=y
+CONFIG_LED_GPIO=y
 # CONFIG_LED_STATUS is not set
 
 #
@@ -1075,7 +1076,11 @@ CONFIG_USB_GADGET_PRODUCT_NUM=0x1010
 # CONFIG_USB_GADGET_DWC2_OTG is not set
 # CONFIG_CI_UDC is not set
 CONFIG_USB_GADGET_VBUS_DRAW=2
-# CONFIG_USB_GADGET_DOWNLOAD is not set
+CONFIG_USB_GADGET_DOWNLOAD=y
+CONFIG_USB_FUNCTION_MASS_STORAGE=y
+# CONFIG_USB_FUNCTION_ROCKUSB is not set
+# CONFIG_USB_FUNCTION_SDP is not set
+# CONFIG_USB_FUNCTION_THOR is not set
 # CONFIG_USB_ETHER is not set
 # CONFIG_USB_HOST_ETHER is not set
 
@@ -1130,7 +1135,7 @@ CONFIG_VIDEO_DT_SIMPLEFB=y
 # CONFIG_FS_BTRFS is not set
 # CONFIG_FS_CBFS is not set
 CONFIG_FS_EXT4=y
-CONFIG_EXT4_WRITE=y
+# CONFIG_EXT4_WRITE is not set
 CONFIG_FS_FAT=y
 CONFIG_FAT_WRITE=y
 CONFIG_FS_FAT_MAX_CLUSTSIZE=65536
diff --git a/configs/nanopi_m1_defconfig b/configs/nanopi_m1_defconfig
index e0ae3c7..6324a71 100644
--- a/configs/nanopi_m1_defconfig
+++ b/configs/nanopi_m1_defconfig
@@ -12,3 +12,4 @@ CONFIG_DEFAULT_DEVICE_TREE="sun8i-h3-nanopi-m1"
 # CONFIG_SPL_EFI_PARTITION is not set
 CONFIG_USB_EHCI_HCD=y
 CONFIG_SYS_USB_EVENT_POLL_VIA_INT_QUEUE=y
+CONFIG_MMC_SUNXI_SLOT_EXTRA=2
diff --git a/configs/qpt-vdih3v1_defconfig b/configs/qpt-vdih3v1_defconfig
new file mode 100644
index 0000000..939c600
--- /dev/null
+++ b/configs/qpt-vdih3v1_defconfig
@@ -0,0 +1,20 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_SPL=y
+CONFIG_MACH_SUN8I_H3=y
+CONFIG_DRAM_CLK=408
+CONFIG_DRAM_ZQ=3881979
+CONFIG_DRAM_ODT_EN=y
+CONFIG_DEFAULT_DEVICE_TREE="sun8i-h3-qpt-vdiv1"
+# CONFIG_SYS_MALLOC_CLEAR_ON_INIT is not set
+# CONFIG_CMD_FLASH is not set
+# CONFIG_SPL_DOS_PARTITION is not set
+# CONFIG_SPL_EFI_PARTITION is not set
+CONFIG_USB_EHCI_HCD=y
+CONFIG_SYS_USB_EVENT_POLL_VIA_INT_QUEUE=y
+CONFIG_MMC_SUNXI_SLOT_EXTRA=2
+CONFIG_LED=y
+CONFIG_LED_BLINK=y
+CONFIG_LED_GPIO=y
+CONFIG_USB_GADGET_DOWNLOAD=y
+CONFIG_USB_FUNCTION_MASS_STORAGE=y
diff --git a/drivers/usb/host/usb-uclass.c b/drivers/usb/host/usb-uclass.c
index 611ea97..2440230 100644
--- a/drivers/usb/host/usb-uclass.c
+++ b/drivers/usb/host/usb-uclass.c
@@ -200,6 +200,7 @@ int usb_stop(void)
 	return err;
 }
 
+extern void qpt_leds_blink(int i);
 static void usb_scan_bus(struct udevice *bus, bool recurse)
 {
 	struct usb_bus_priv *priv;
@@ -217,8 +218,10 @@ static void usb_scan_bus(struct udevice *bus, bool recurse)
 		printf("failed, error %d\n", ret);
 	else if (priv->next_addr == 0)
 		printf("No USB Device found\n");
-	else
+	else {
+		qpt_leds_blink(bus->seq);
 		printf("%d USB Device(s) found\n", priv->next_addr);
+	}
 }
 
 static void remove_inactive_children(struct uclass *uc, struct udevice *bus)
diff --git a/drivers/video/dw_hdmi.c b/drivers/video/dw_hdmi.c
index dbad0e2..7fdd1f3 100644
--- a/drivers/video/dw_hdmi.c
+++ b/drivers/video/dw_hdmi.c
@@ -696,7 +696,7 @@ int dw_hdmi_read_edid(struct dw_hdmi *hdmi, u8 *buf, int buf_size)
 	} else {
 		ret = hdmi_read_edid(hdmi, 0, buf);
 		if (ret) {
-			debug("failed to read edid.\n");
+			printf("%s: Failed to read edid.\n", __func__);
 			return -1;
 		}
 
@@ -706,6 +706,7 @@ int dw_hdmi_read_edid(struct dw_hdmi *hdmi, u8 *buf, int buf_size)
 		}
 	}
 
+	edid_print_info((struct edid1_info *)buf);
 	return edid_size;
 }
 
diff --git a/env/env.c b/env/env.c
index 5c0842a..fd19753 100644
--- a/env/env.c
+++ b/env/env.c
@@ -5,6 +5,7 @@
  */
 
 #include <common.h>
+#include <autoboot.h>
 #include <environment.h>
 
 DECLARE_GLOBAL_DATA_PTR;
@@ -185,6 +186,9 @@ int env_load(void)
 	struct env_driver *drv;
 	int prio;
 
+	/* ROY: add GPIO-reset default hw-version compact here */
+	emc_check_hw_support();
+
 	for (prio = 0; (drv = env_driver_lookup(ENVOP_LOAD, prio)); prio++) {
 		int ret;
 
diff --git a/include/autoboot.h b/include/autoboot.h
index 3ebd6f9..9f81d0c 100644
--- a/include/autoboot.h
+++ b/include/autoboot.h
@@ -11,6 +11,9 @@
 #ifndef __AUTOBOOT_H
 #define __AUTOBOOT_H
 
+void emc_check_hw_support(void);
+void qpt_leds_blink(int i);
+
 #ifdef CONFIG_AUTOBOOT
 /**
  * bootdelay_process() - process the bootd delay
diff --git a/include/config_distro_bootcmd.h b/include/config_distro_bootcmd.h
index f2d02fe..643188d 100644
--- a/include/config_distro_bootcmd.h
+++ b/include/config_distro_bootcmd.h
@@ -337,11 +337,10 @@
 	BOOTENV_SHARED_EFI \
 	"boot_prefixes=/ /boot/\0" \
 	"splashpos=m,m\0" \
-	"splashimage=66000000\0" \
-	"loadsplash= " \
-		"for prefix in ${boot_prefixes}; do " \
-			"if test -e mmc 0 ${prefix}boot.bmp; then " \
-				"load mmc 0 ${splashimage} ${prefix}boot.bmp; " \
+	"splashimage=0x66000000\0" \
+	"loadsplash=for prefix in ${boot_prefixes}; do " \
+			"if test -e mmc 1 ${prefix}boot.bmp; then " \
+				"load mmc 1 ${splashimage} ${prefix}boot.bmp; " \
 				"bmp d ${splashimage}; " \
 			"fi; " \
 		"done\0" \
@@ -403,6 +402,7 @@
 	BOOT_TARGET_DEVICES(BOOTENV_DEV)                                  \
 	\
 	"distro_bootcmd=" BOOTENV_SET_SCSI_NEED_INIT                      \
+		"run emergency_boot && source ${scriptaddr};" \
 		"for target in ${boot_targets}; do "                      \
 			"run bootcmd_${target}; "                         \
 		"done\0"
diff --git a/include/configs/sunxi-common.h b/include/configs/sunxi-common.h
index 042d506..7863ddc 100644
--- a/include/configs/sunxi-common.h
+++ b/include/configs/sunxi-common.h
@@ -470,7 +470,7 @@ extern int soft_i2c_gpio_scl;
 #elif CONFIG_DM_VIDEO
 #define CONFIG_SYS_WHITE_ON_BLACK
 #define CONSOLE_STDOUT_SETTINGS \
-	"stdout=serial,vidconsole\0" \
+	"stdout=serial\0" \
 	"stderr=serial,vidconsole\0"
 #else
 #define CONSOLE_STDOUT_SETTINGS \
diff --git a/make.sh b/make.sh
new file mode 100755
index 0000000..305aef5
--- /dev/null
+++ b/make.sh
@@ -0,0 +1,40 @@
+#!/bin/bash
+
+build_target()
+{
+    # ENV
+    which arm-linux-gnueabihf-gcc 2>/dev/null || {
+        echo "需要设置环境变量, 指定交叉编译工具链: arm-linux-gnueabihf- 的路径, 例如:"
+        echo 'export PATH=$PATH:/xxx/gcc-linaro-7.2.1-2017.11-x86_64_arm-linux-gnueabihf/bin'
+        exit -1
+    }
+    # BUILD
+    make CROSS_COMPILE=arm-linux-gnueabihf- || exit -1
+}
+
+patch_src()
+{
+    git diff sun8i-patched > z-1000-QPT-uboot.patch
+}
+
+usage()
+{
+    echo "./make.sh <build> --- 编译uboot"
+    echo "          <patch> --- 生成armbian补丁, 需要手动拷贝到: userpatches/u-boot/u-boot-sunxi/branch_default/"
+    echo ""
+    echo "  注意: 若修改.config默认选项, 需要同步更新armbian/config相关配置"
+}
+
+case $1 in
+    patch)
+        echo "Patch code modify from armbian-original."
+        patch_src
+    ;;
+    build)
+        echo "Build u-boot-sunxi-with-spl.bin"
+        build_target
+    ;;
+    *)
+        usage
+    ;;
+esac
\ No newline at end of file
